astropy/utils/console.py

```diff
@@ -49,42 +50,33 @@ class _IPython:
 
     @classproperty
     def get_ipython(cls):
-        try:
+        if HAS_IPYTHON:
             from IPython import get_ipython
-        except ImportError:
-            pass
-        return get_ipython
+
+            return get_ipython
+        raise ModuleNotFoundError("IPython is not installed")
 
     @classproperty
     def OutStream(cls):
         if not hasattr(cls, "_OutStream"):
-            cls._OutStream = None
-            try:
-                cls.get_ipython()
-            except NameError:
```

Comment:
I don't know what this NameError was for but looks like handling it has been dropped in this PR. Can you please clarify? Thanks!

Reply:
The `get_ipython` method was oddly written in such a way that its return variable was only bound if IPython was installed, meaning it was effectively relying on a NameError being raised otherwise. Since I also rewrote the method in a more conventional manner, this handling isn't needed any more

Reply:
Still, looking at code in `main`, the original design looks like it never intended to throw exception if IPython is not installed, but this PR does. Can we refactor in a way that does not throw exception? Thanks for your patience!

Reply:
I'm confused

> original design looks like it never intended to throw exception if IPython is not installed

if you mean the `get_ipython` class method, it seems to me that it *was* actually designed to throw an exception since that's how it's been used through `except NameError` (albeit a quirky exception to catch in your own code).

> but this PR does

I don't understand, where else would an exception be raised in my version ?

Reply:
Actually the `NameError` on `main` makes no sense to me. When I try to emulate the implementation, I get `UnboundLocalError`. But that is neither here nor there...

The point I was trying to make is that `get_ipython` on `main` defers the exception handling to `OutStream`. But in this PR, you explicitly throws `ModuleNotFoundError` from within `get_ipython` if IPython is not installed. I think that is a subtle behavior change?

Reply:
`UnboundLocalError` is a subclass of `NameError`, so the code on main is equivalent to
```python
try:
    from IPython import get_ipython
    return get_ipython
except ImportError:
    raise NameError from None
```

Reply:
In trying to understand this part of the code more, I get even more confused. I see this:

https://github.com/astropy/astropy/blob/cec24e83efea38ee18b80b7b34693a209d07d595/astropy/logger.py#L310

And this:

https://github.com/astropy/astropy/blob/cec24e83efea38ee18b80b7b34693a209d07d595/astropy/conftest.py#L150-L153

But `get_ipython` isn't imported from anywhere, but looks like the code is expected `NameError` like the one in `console.py`. 🤯

Reply:
oh wow, I get the confusion !
After inspection, I think it's actually unrelated : this pattern seems to be used as a cheap proxy to check whether the interpreter is currently being run through `ipython`, as one of the expected side effects is that you get `get_ipython` function defined in the global scope.

```
❯ ipython
Python 3.12.2 (main, Feb 22 2024, 09:53:39) [Clang 15.0.0 (clang-1500.1.0.2.5)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.23.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: get_ipython()
Out[1]: <IPython.terminal.interactiveshell.TerminalInteractiveShell at 0x105e97950>
```

This pattern is used only in a couple places, I could easily refactor and/or comment them all to avoid future confusion. What do you think ? Should I do it as part of this PR or would you rather I open a new one ?

Reply:
actually it was easy enough that I just went for it: https://github.com/astropy/astropy/pull/16369

Reply:
Also, I figured that this `_IPython.get_ipython` method was not actually needed, so I refactored it. To make it easier for you to follow after I rebased, I've kept this one change as a separate commit (that should eventually get squashed).

Reply:
Thanks, does look much cleaner now. But I still don't know what is the point of this whole `_IPython` class now that I dig into the code. I can't seem to see it being used anywhere and it is "hidden".

Anyways, should we get https://github.com/astropy/astropy/pull/16369 in first?

Reply:
Ah nvm, found it. Just used internally here for `isatty`.

Reply:
>  we get https://github.com/astropy/astropy/pull/16369 in first?

these PRs are actually completely independent as it turns out; they can be merged in any order